many: install bash completion files in writable directory

valentindavid opened the pull request:
Snapd installs bash completion files from snaps in
`/usr/share/bash-completion/completions` which in some distributions
is a read-only filesystem.

Instead of installing them in `/usr` we can install them within
`/var/lib/snapd` which should always be writable.

To enable that path in bash completions we only need to add the path
to `XDG_DATA_DIRS`.

An alternative could be to install the file in
`/etc/bash_completion.d`.

~~DRAFT: This is incomplete as we need to migrate symlinks in
`/usr/share/bash-completion/completions` to `/var/lib/snapd`~~
Migration of the path was added.

anonymouse64 reviewed (commented):
thanks for working on this, were you able to test this out on an Ubuntu Core system to ensure that we don't also need anything else like extra packages in the base snap too? There was some discussion about missing packages a while ago in https://github.com/snapcore/core18/issues/89, not sure if any of that work ever happened (or maybe this was triggered by working on that bug?)

anonymouse64 commented:
Ah I see you also opened a PR to the core20 repo, sorry I didn't see a link in the description and only noticed that this PR is mentioned there.

valentindavid commented on data/env/snapd.sh.in:
```diff
@@ -20,3 +20,7 @@ if [ -n "${XDG_DATA_DIRS##*${snap_xdg_path}}" ] && [ -n "${XDG_DATA_DIRS##*${sna
+snap_data_path="/var/lib/snapd/share"
+if [ -n "${XDG_DATA_DIRS##*${snap_data_path}}" ] && [ -n "${XDG_DATA_DIRS##*${snap_data_path}:*}" ]; then
+    export XDG_DATA_DIRS="${XDG_DATA_DIRS}:${snap_data_path}"
+fi
```
Did some changes there. Also made the regular expressions a bit more precise.

valentindavid commented on data/completion/bash/complete.sh:
```diff
@@ -102,10 +102,10 @@ _complete_from_snap() {
 
 # this file can be sourced directly as e.g. /usr/lib/snapd/complete.sh, or via
-# a symlink from /usr/share/bash-completion/completions/. In the first case we
+# a symlink from /var/lib/snapdshare/bash-completion/completions/. In the first case we
```
Fixed

valentindavid commented:
> thanks for working on this, were you able to test this out on an Ubuntu Core system to ensure that we don't also need anything else like extra packages in the base snap too? There was some discussion about missing packages a while ago in [snapcore/core18#89](https://github.com/snapcore/core18/issues/89), not sure if any of that work ever happened (or maybe this was triggered by working on that bug?)



> thanks for working on this, were you able to test this out on an Ubuntu Core system to ensure that we don't also need anything else like extra packages in the base snap too? There was some discussion about missing packages a while ago in [snapcore/core18#89](https://github.com/snapcore/core18/issues/89), not sure if any of that work ever happened (or maybe this was triggered by working on that bug?)

So, I have tested it with https://github.com/snapcore/core20/pull/121
I have tested it by installing docker and trying to use completion on a docker command line.

alfonsosanchezbeato commented on overlord/snapstate/backend/aliases.go:
```diff
@@ -46,12 +46,13 @@ func (b Backend) UpdateAliases(add []*Alias, remove []*Alias) error {
+		for _, completersPath := range []string{dirs.CompletersDir, dirs.LegacyCompletersDir} {
+			completer := filepath.Join(completersPath, alias.Name)
+			target, err := os.Readlink(completer)
+			if err == nil || target == alias.Target {
```
If you change the logic here to remove inside the `if` you need a also to change the `||` to `&&`. Anyway, I think that the coding conventions in snapd prefer in general to use `continue` when some condition inside the loop is not met.

valentindavid replied:
Oops. Thank you for spotting that.

anonymouse64 commented:
@valentindavid can you merge master or rebase on master to trigger the tests to be run again? I fixed the PR title which should fix the tests here

valentindavid commented:
> @valentindavid can you merge master or rebase on master to trigger the tests to be run again? I fixed the PR title which should fix the tests here

@anonymouse64 I have rebased

bboozzoo commented on dirs/dirs.go:
```diff
@@ -457,7 +458,8 @@ func SetRootDir(rootdir string) {
 	CompletionHelperInCore = filepath.Join(CoreLibExecDir, "etelpmoc.sh")
-	CompletersDir = filepath.Join(rootdir, "/usr/share/bash-completion/completions/")
+	LegacyCompletersDir = filepath.Join(rootdir, "/usr/share/bash-completion/completions/")
+	CompletersDir = filepath.Join(rootdir, snappyDir, "share/bash-completion/completions/")
```
I wonder if `/var/lib/snapd/desktop/bas-completion/completions` isn't more appropriate here which has an upside of already being included in XDG_DATA_DIRS

valentindavid replied:
My problem with it is `desktop`. I do not really like to keep a name for historical reason. And I would prefer to migrate files from `/var/lib/snapd/desktop` to `/var/lib/snapd/share`.

valentindavid replied:
It now uses `/var/lib/snapd/desktop`.

bboozzoo reviewed (commented):
Some questions, also needs a merge from master or a rebase

valentindavid commented on tests/main/user-session-env/task.yaml:
```diff
@@ -38,20 +38,22 @@ execute: |
             if [ -d /usr/lib/systemd/user-environment-generators ]; then
-                MATCH 'XDG_DATA_DIRS=.*[:]?/var/lib/snapd/desktop[:]?.*' < "${user}-session-env"
-                MATCH "PATH=.*[:]?${SNAP_MOUNT_DIR}/bin[:]?.*" < "${user}-session-env"
+                MATCH 'XDG_DATA_DIRS=(^|.*[:])/var/lib/snapd/desktop($|[:].*)' < "${user}-session-env"
```
Yes. I have rewritten the regex. They did not make much sense with "^" not at the beginning of the pattern anyway.

valentindavid replied:
This has been reverted

valentindavid commented on cmd/snap/gnupg2_test.go:
```diff
@@ -24,4 +24,4 @@ import (
 
 // FIXME: drop once gpg2 is the default
-var _ = Suite(&SnapKeysSuite{GnupgCmd: "/usr/bin/gpg2"})
+var _ = Suite(&SnapKeysSuite{GnupgCmd: "/usr/bin/gpg"})
```
Removed.

mardy reviewed (commented):
If I understand correctly, this is adding a directory to `XDG_DATA_DIRS` for the whole system.

But then, should we also add this directory to the AppArmor profile of all snaps (either as allowed or even denied), that otherwise might try to look up this directory when loading their data files, and then spam the journal with AppArmor warnings?

valentindavid commented:
OK, adding a path in `XDG_DATA_DIRS` seemed to be a problem. So I just changed int install files in `/var/lib/snapd/desktop` which already exists.

valentindavid commented:
There is some issue with bash-completion 2.1 and prior (ubuntu 14.04 and prior). Trying to work around that.

bboozzoo commented on image/preseed/reset.go:
```diff
@@ -135,37 +135,39 @@ func ResetPreseededChroot(preseedChroot string) error {
-	for _, fileInfo := range files {
-		if fileInfo.Mode()&os.ModeSymlink == 0 {
-			continue
+	for _, completersPath := range []string{dirs.CompletersDir, dirs.LegacyCompletersDir} {
```
Hmm perhaps @stolowski has some insight here, but AFAIU this is called when cleaning up a system image when it was preseeded by the current snapd version. Much of the trouble before was due to snapd writing under the real /usr/share/bas-completion, which we don't do now. Thus, if the current version only writes all the files files under dirs.CompletersDir now, the whole process becomes much simpler now as that directory will already get removed, and there's nothing to do?

stolowski replied:
Yes, I think (AFAIU) we can just add `CompletersDir` to the list of directories removed recursively (leaving the top dir intact) above; I think though that we need to keep existing logic to clean `dirs.LegacyCompletersDir` in case a new version of `snap-preseed` command is run against older snapd?

bboozzoo commented on wrappers/binaries.go:
```diff
@@ -21,13 +21,57 @@
+		return true, false
+	}
+	defer fd.Close()
+	buf := new(strings.Builder)
```
```suggestion
	buf := &bytes.Buffer{}
```
unless you have a specific use for the builder, which I don't think there is one here

valentindavid replied:
This line has been removed.

bboozzoo commented on wrappers/binaries.go:
```diff
@@ -21,13 +21,57 @@
+	defer fd.Close()
+	buf := new(strings.Builder)
+	io.Copy(buf, fd)
+	bash_completion := buf.String()
```
hmm, so we load all of the file to a buffer which then returns it at a string, this could be done by:
```
d, err := ioutil.ReadFile(dirs.BasCompletionScript)
// check err
bash_completion := string(d)
```
but then we look a match inside the file, so perhaps we don't really need to load all 76kB of that file, but rather read it line by line looking for a match?

```
reRelease := regexp.MustCompile(`(?:^|\n)# *RELEASE: ([0-9]+)[.]([0-9]+)(?:$|\n)`)
reBashCompletionVersinfo := regexp.MustCompile(`(?:^|\n)BASH_COMPLETION_VERSINFO=`)
s := bufio.NewScanner(fd)
for s.Scan() {
    line := s.Text()
    // process
}
if err := s.Err(); err != nil {
    return fmt.Errorf("cannot read file: %v", err)
}
```

valentindavid replied:
Changed use the line scanner.

bboozzoo commented on wrappers/binaries.go:
```diff
@@ -21,13 +21,57 @@
+	io.Copy(buf, fd)
+	bash_completion := buf.String()
+
+	re := regexp.MustCompile(`(?:^|\n)# *RELEASE: ([0-9]+)[.]([0-9]+)(?:$|\n)`)
```
I don't see `RELASE:` in that file on Arch, was is present in some older releases?

bboozzoo replied:
Do you know which in bash version it was replaced with BASH_COMPLETION_VERSINFO? maybe it's so old that we don't need to care anymore?

valentindavid replied:
2.6. Xenial has 2.1 so we need to match `RELEASE:`. If it has `BASH_COMPLETION_VERSINFO` then it supports `XDG_DATA_DIRS` since that support was added in 2.2 which is prior to using `BASH_COMPLETION_VERSINFO`.

valentindavid replied:
Also it is bash-completion. The version of bash is irrelevant.

valentindavid replied:
I had changed the code to search patterns line by line like you suggested. This makes the order of checks irrelevant for performance.

bboozzoo commented on wrappers/binaries.go:
```diff
@@ -21,13 +21,57 @@
 	"github.com/snapcore/snapd/snap"
 )
 
+func detectLegacy() (noCompletion bool, legacyCompletion bool) {
```
could use a doc commend describing the reason to run this detection, eg.
```
// detectLegacy check whether the bash completion support is recent enough
// such that the completion files can be placed under an arbitrary path listed in
// XDG_DATA_DIRS.
```

valentindavid replied:
Done

mardy commented on wrappers/binaries.go:
```diff
@@ -21,13 +21,71 @@
+// detectLegacy checks whether bash-completion is version 2.2 or later.
+// Prior to 2.2, bash-completion does not support XDG_DATA_DIRS.
+// If it fails to detect the version of bash-completion, it disables
+// completion.
```
It might be easier if this function returned an enum,

    NoCompletion,
    LegacyCompletion,
    FullCompletion,

valentindavid replied:
Done

mardy commented on wrappers/binaries.go:
```diff
@@ -45,7 +103,16 @@ func AddSnapBinaries(s *snap.Info) (err error) {
 
 	completeSh := dirs.CompleteShPath(s.Base)
-	noCompletion := !osutil.IsWritable(dirs.CompletersDir) || !osutil.FileExists(completeSh)
+	noCompletion := !osutil.FileExists(completeSh)
```
it sounds like this could be moved inside detectLegacy() (if we rename it to detectCompletionSupport and make it return an enum)

valentindavid replied:
Done

mardy reviewed (approved):
Thanks! Looks good, I have a few suggestions but nothing critical. :-)

bboozzoo commented on overlord/snapstate/backend/aliases.go:
```diff
@@ -46,12 +46,13 @@ func (b Backend) UpdateAliases(add []*Alias, remove []*Alias) error {
+			completer := filepath.Join(completersPath, alias.Name)
+			target, err := os.Readlink(completer)
+			if err == nil && target == alias.Target {
+				os.Remove(completer)
```
it would have to be a symlink which points to location under /snap/bin (or /var/lib/snapd/snap/bin) which is unlikely

valentindavid replied:
What @bboozzoo says. If `target == alias.Target` then it is better to remove it.

bboozzoo commented on tests/completion/lib.exp0:
```diff
@@ -66,5 +66,5 @@ proc brexit {} {
 spawn env -u PS1 bash --norc -i
 next
-send ". /usr/share/bash-completion/bash_completion; . ../../../data/completion/snap\n"
+send ". /usr/share/bash-completion/bash_completion; . ../../../data/completion/bash/snap\n"
```
hmm so how did this even work before? perhaps we should have another:
```
send "set -e\n"
next
```
to ensure that errors result in failure?

valentindavid replied:
I think the reason that it was working is that the file was sourced again in the .exp files that use `lib.exp0`.

I am afraid that `set -e` will have unexpected effects in an interactive shell, because it is not a common thing to do. It is meant for non-interactive scripts. I do not want to debug that. Feel free to open a separate pull request.

valentindavid commented on wrappers/binaries.go:
```diff
@@ -21,13 +21,71 @@
+		return true, false
+	}
+
+	if major < 2 {
+		return false, true
+	}
+
+	if major > 2 {
+		return false, false
+	}
+
+	return false, minor < 2
```
In practice, because 3.1 will not reach this part of the code, that would work. Although not in theory. But this is not how we compare versions. Any time someone would read that part of the code would see `major >= 2 && minor >= 2`, they would think there is a bug, until they figure out that major cannot be more than 2, in that part of the code... in practice, but not in theory.

bboozzoo replied:
Well, yes code is structured in a way that for 2.6+ (or at least any version that has BASH_COMPLETION_VERSINFO=) we just return early and the rest isn't even reached, so the check `if major > 2` is actually impossible to reach, isn't it?
Perhaps there should be a comment after the for loop indicating that  the rest is reached only if the bash-completion had the old string and now we need to check the version which can be up to 2.5 really. FWIW if any future version drops the line we match on, the code will be incorrect anyway.

valentindavid replied:
I will try to rewrite the code a different way. @mardy also had comments on this function anyway. But I cannot write `major >= 2 && minor >= 2`, that burns my eyes.

valentindavid replied:
I have changed to code to use `strutil.CompareVersion`.

valentindavid commented on tests/completion/snippets/task.yaml:
```diff
@@ -20,4 +20,5 @@ execute: |
   #shellcheck disable=SC1090
   source "${SPREAD_PATH}/${SPREAD_SUITE}/${SPREAD_VARIANT}.vars"
   export _OUT0 _OUT1 _OUT2 _KEY1 _KEY2 _COMP
+  export XDG_DATA_DIRS="${XDG_DATA_DIRS:-}${XDG_DATA_DIRS:+:}/var/lib/snapd/desktop"
```
`sudo ... -E -u test ...` is used here. sudo documentation claims that it would just fail if `-E` was not allowed. Unless I am not understanding correctly. Am I missing something?

bboozzoo replied:
Ah, I missed -E. It's fine then.

valentindavid commented on tests/completion/snippets/task.yaml:
```diff
@@ -20,4 +20,5 @@ execute: |
   source "${SPREAD_PATH}/${SPREAD_SUITE}/${SPREAD_VARIANT}.vars"
   export _OUT0 _OUT1 _OUT2 _KEY1 _KEY2 _COMP
+  export XDG_DATA_DIRS="${XDG_DATA_DIRS:-}${XDG_DATA_DIRS:+:}/var/lib/snapd/desktop"
   sudo PATH="$PATH" -E -u test expect -d -f "$d"/task.exp
```
The tests use `expect` which is not really possible to use on Ubuntu Core. I will look into writing some tests without using `expect` on Wednesday.

bboozzoo replied:
There are people who use snapd on Yocto, where read only rootfs is quite commonly used. OTOH we haven't had any reports where completions would break such installs. I guess it has to do with mostly command line tools shipping completions (if at all), while those systems, much like Ubuntu Core are not really intended to be accessed directly though ssh or similar. So I'd say it's ok to address this in a followup, though with the new completions location it shouldn't be a problem at all (unless they ship old bash-completions but then we can always push back).

valentindavid replied:
There is now `tests/core/bash-completion` which tests completion on UC22.

valentindavid commented on image/preseed/reset.go:
```diff
@@ -135,37 +135,39 @@ func ResetPreseededChroot(preseedChroot string) error {
 	// e.g.
 	// lxd.lxc -> /snap/core/current/usr/lib/snapd/complete.sh
 	// lxc -> lxd.lxc
-	files, err := ioutil.ReadDir(filepath.Join(preseedChroot, dirs.CompletersDir))
```
Done

mardy commented on wrappers/binaries.go:
```diff
@@ -60,11 +135,28 @@ func AddSnapBinaries(s *snap.Info) (err error) {
 			continue
 		}
 		// symlink the completion snippet
+		compDir := dirs.CompletersDir
+		if completion == legacyCompletion {
+			compDir = dirs.LegacyCompletersDir
+		}
+		if err := os.MkdirAll(compDir, 0755); err != nil {
+			return err
+		}
```
nitpick: unless there is some harm in doing this, I'd move these lines outside of the `for` loop.

valentindavid replied:
We might not get to that point. For example if `app.Completer == ""` for all apps.

mardy commented on image/preseed/reset.go:
```diff
@@ -31,6 +31,50 @@ import (
+// lxd.lxc -> /snap/core/current/usr/lib/snapd/complete.sh
+// lxc -> lxd.lxc
+func resetCompletionSymlinks(completersPath string, preseedChroot string) error {
+	files, err := ioutil.ReadDir(filepath.Join(preseedChroot, completersPath))
```
nitpick: you are joining `preseedChroot` and `completersPath` below, too, so it would make sense to do it once here and remember the result in a variable.

valentindavid replied:
Done as @bboozzoo suggested.

mardy commented on image/preseed/reset.go:
```diff
@@ -31,6 +31,50 @@ import (
+func resetCompletionSymlinks(completersPath string, preseedChroot string) error {
+	files, err := ioutil.ReadDir(filepath.Join(preseedChroot, completersPath))
+	if err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("error reading %s: %v", completersPath, err)
```
here it might make sense to print the path with the preseedChroot prefix as well

valentindavid replied:
Done

mardy commented on image/preseed/reset.go:
```diff
@@ -31,6 +31,50 @@ import (
+		if err != nil {
+			return fmt.Errorf("error reading symlink target of %s: %v", other, err)
+		}
+		if _, ok := completeShSymlinks[target]; ok {
+			if err := os.Remove(other); err != nil {
+				return fmt.Errorf("error removing symlink %s: %v", other, err)
+			}
+		}
```
Just wondering: would it make sense to just check if the target exists, and remove the current symlink if the target does not exist?
It would be slightly less efficient, but then we wouldn't need to worry about symlinks like `"./file-in-same-dir"`. Though I'm probably inventing problems here. :-)

valentindavid replied:
This is not my code. This was moved for this PR, but this part was not modified. I would prefer not to change it.

The way symlinks are created is:
 - one layer of symlinks point to some `complete.sh` from somewhere else (within snapd snap most of the time).
 - for every alias, there is a symlink pointing to the main symlink.

So to me this code looks like a correct way to remove symlinks created by snapd.

I am not totally sure if that applies to preseeding, but I do not think we should remove invalid symlinks not created by snapd.

mardy reviewed (approved):
LGTM, thanks for the updates!

I got a few nitpick comments inside, but unless you need to do more changes, you can safely ignore them.

bboozzoo commented on wrappers/binaries.go:
```diff
@@ -21,13 +21,88 @@
+	}
+	defer fd.Close()
+
+        // Up to 2.5
+	releaseOld := regexp.MustCompile(`^# *RELEASE: ([0-9.]+)$`)
+        // 2.6 and later
+	releaseNew := regexp.MustCompile(`^BASH_COMPLETION_VERSINFO=`)
```
tabs vs spaces

```suggestion
	// Up to 2.5
	releaseOld := regexp.MustCompile(`^# *RELEASE: ([0-9.]+)$`)
	// 2.6 and later
	releaseNew := regexp.MustCompile(`^BASH_COMPLETION_VERSINFO=`)
```

valentindavid replied:
Fixed.

bboozzoo reviewed (approved):
LGTM, little tweaks but nothing major

valentindavid commented on tests/core/bash-completion/test-completion.py:
```diff
@@ -0,0 +1,118 @@
+        os.close(tmp_fd)
+
+    @classmethod
+    async def create(cls, pty, init):
```
Done

valentindavid commented on tests/core/bash-completion/test-completion.py:
```diff
@@ -0,0 +1,118 @@
+
+class Shell:
+    @classmethod
+    def prepare_tty(cls):
```
Done

valentindavid commented on tests/core/bash-completion/test-completion.py:
```diff
@@ -0,0 +1,118 @@
+
+async def run(init, executable):
+    with contextlib.closing(Pty()) as pty:
+        async with contextlib.aclosing(await Shell.create(pty, init)) as shell:
```
Done

valentindavid commented on image/preseed/reset.go:
```diff
@@ -130,43 +174,9 @@ func ResetPreseededChroot(preseedChroot string) error {
-				return fmt.Errorf("error removing symlink %s: %v", other, err)
-			}
+	for _, completersPath := range []string{dirs.CompletersDir, dirs.LegacyCompletersDir} {
+		if err := resetCompletionSymlinks(completersPath, preseedChroot); err != nil {
```
Done
